# 1. Introduction to Prompt A/B Testing

A/B testing compares two or more prompt variations to determine which performs better.

## Why Use A/B Testing?
- Improve response quality
- Increase accuracy
- Reduce hallucination
- Improve consistency
- Optimize prompt engineering strategy

## Example
Prompt A:
"Summarize the following text."

Prompt B:
"Summarize the following text in 3 bullet points with key insights."

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

In [2]:
prompt_A = """
Summarize the following customer review.
"""

prompt_B = """
Summarize the following customer review in 3 concise bullet points.
Mention positives and negatives separately.
"""

print("Prompt A:\n", prompt_A)
print("\nPrompt B:\n", prompt_B)

Prompt A:
 
Summarize the following customer review.


Prompt B:
 
Summarize the following customer review in 3 concise bullet points.
Mention positives and negatives separately.



In [3]:
test_inputs = [
    "The phone camera is excellent but battery drains quickly.",
    "Laptop performance is great for gaming, but it overheats.",
    "The headphones have amazing sound quality and comfortable fit."
]

dataset = pd.DataFrame({
    "input_text": test_inputs
})

dataset

,input_text
0,The phone camera is excellent but battery drai...
1,"Laptop performance is great for gaming, but it..."
2,The headphones have amazing sound quality and ...


# 2. Controlled Test Scenarios

To compare prompts fairly, control:

- Same model
- Same temperature
- Same dataset
- Same token limits
- Same evaluation criteria

This ensures prompt is the only changing variable.

In [4]:
responses_A = [
    "Good camera but poor battery.",
    "Great gaming performance but overheating issue.",
    "Excellent sound and comfort."
]

responses_B = [
    """Positives:
- Excellent camera

Negatives:
- Battery drains quickly""",

    """Positives:
- Strong gaming performance

Negatives:
- Overheats during use""",

    """Positives:
- Great sound quality
- Comfortable fit

Negatives:
- No major issues"""
]

# 3. Evaluation Criteria

Each response is scored from 1–5.

## Metrics
1. Relevance
2. Accuracy
3. Completeness
4. Clarity
5. Consistency

Total Score = Sum of all metrics
Maximum = 25

In [5]:
def score_response(relevance, accuracy, completeness, clarity, consistency):
    total = relevance + accuracy + completeness + clarity + consistency
    return total

In [6]:
scores_A = [
    score_response(4,4,3,4,4),
    score_response(4,4,3,4,4),
    score_response(4,4,3,4,4)
]

scores_B = [
    score_response(5,5,5,5,5),
    score_response(5,5,5,5,5),
    score_response(5,5,4,5,5)
]

In [7]:
results = pd.DataFrame({
    "Input": test_inputs,
    "Score_A": scores_A,
    "Score_B": scores_B
})

results


,Input,Score_A,Score_B
0,The phone camera is excellent but battery drai...,19,25
1,"Laptop performance is great for gaming, but it...",19,25
2,The headphones have amazing sound quality and ...,19,24


In [8]:
avg_A = np.mean(scores_A)
avg_B = np.mean(scores_B)

print("Average Score Prompt A:", avg_A)
print("Average Score Prompt B:", avg_B)

Average Score Prompt A: 19.0
Average Score Prompt B: 24.666666666666668


In [9]:
consistency_A = np.std(scores_A)
consistency_B = np.std(scores_B)

print("Prompt A Consistency (lower is better):", consistency_A)
print("Prompt B Consistency (lower is better):", consistency_B)

Prompt A Consistency (lower is better): 0.0
Prompt B Consistency (lower is better): 0.4714045207910317


# 4. Robustness Testing

Robust prompts should perform well across:

- Short inputs
- Long inputs
- Noisy inputs
- Ambiguous inputs
- Edge cases

In [10]:
robust_inputs = [
    "Bad.",
    "This product exceeded expectations in every way.",
    "Average product but support was terrible."
]

robust_inputs

['Bad.',
 'This product exceeded expectations in every way.',
 'Average product but support was terrible.']

In [11]:
experiment_log = pd.DataFrame({
    "Experiment_ID": [1],
    "Date": [datetime.now()],
    "Prompt_Avg_Score": [avg_A],
    "Prompt_B_Avg_Score": [avg_B],
    "Winner": ["Prompt B"]
})

experiment_log

,Experiment_ID,Date,Prompt_Avg_Score,Prompt_B_Avg_Score,Winner
0,1,2026-06-22 12:32:49.082624,19.0,24.666667,Prompt B


In [12]:
if avg_B > avg_A:
    winner = "Prompt B"
else:
    winner = "Prompt A"

print("Best Performing Prompt:", winner)

Best Performing Prompt: Prompt B


# Prompt A/B Testing Comparison Report

## Experiment Summary
Two prompt versions were tested on multiple customer reviews.

## Prompt A
Simple summarization prompt.

## Prompt B
Structured summarization prompt with bullet formatting and sentiment separation.

## Results
- Prompt A Average Score: 19
- Prompt B Average Score: 24.67

## Findings
Prompt B performed better because:
- More structured outputs
- Better completeness
- Higher clarity
- Better consistency

## Best Prompt Strategy
Use structured prompts with explicit output formatting instructions.

## Conclusion
Prompt engineering significantly affects LLM output quality.
A/B testing helps identify the most effective prompting strategy.